Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_PlayerClimb.cs

```diff
@@ -0,0 +1,188 @@
+            if (collisionFade == null)
+            {
+                collisionFade = headCamera.gameObject.AddComponent<VRTK_HeadsetCollisionFade>();
+            }
+            collisionFade.HeadsetCollisionDetect += new HeadsetCollisionEventHandler(OnHeadsetCollisionDetected);
+            collisionFade.HeadsetCollisionEnded += new HeadsetCollisionEventHandler(OnHeadsetCollisionEnded);
+        }
+
+        private Vector3 GetPosition(Transform objTransform)
+        {
+            if (usePlayerScale)
+            {
+                return Vector3.Scale(objTransform.localPosition, gameObject.transform.localScale);
+            }
+
+            return objTransform.localPosition;
+        }
+
+    
+        private void OnGrabObject(object sender, ObjectInteractEventArgs e)
+        {
+            VRTK_InteractableObject interactObject = e.target.GetComponent<VRTK_InteractableObject>();
+            if( interactObject && interactObject.AttachIsClimbObject() )
+            {
+                if (useGravity)
+                {
+                    playerPhysics.StopPhysicsFall();
+                }
+
+                OnPlayerClimbStarted(SetPlayerClimbEvent(e.controllerIndex, e.target));
+                isClimbing = true;
+                controllerTransform = ((VRTK_InteractGrab)sender).gameObject.transform;
+                startControllerPosition = GetPosition(controllerTransform);
+                startPosition = gameObject.transform.position;
+            }
+        }
+
+        private void OnUngrabObject(object sender, ObjectInteractEventArgs e)
+        {
+            VRTK_InteractableObject interactObject = e.target.GetComponent<VRTK_InteractableObject>();
+            if (interactObject && interactObject.AttachIsClimbObject() && 
+                IsActiveClimbingController(((VRTK_InteractGrab)sender).gameObject))
+            {
+                Ungrab(true, e.controllerIndex, e.target);
+            }
+        }
+
+        private void OnTeleport(object sender, DestinationMarkerEventArgs e)
+        {
+            Ungrab(false, e.controllerIndex, e.target.gameObject);
+        }
+
+        private void OnHeadsetCollisionDetected(object sender, HeadsetCollisionEventArgs e)
+        {
+            headsetColliding = true;
+        }
+
+        private void OnHeadsetCollisionEnded(object sender, HeadsetCollisionEventArgs e)
+        {
+            headsetColliding = false;
+        }
+
+        private void Ungrab(bool carryMomentum, uint controllerIndex, GameObject target)
+        {
+            OnPlayerClimbEnded(SetPlayerClimbEvent(controllerIndex, target));
+            isClimbing = false;
+
+            // Move to the last safe spot
+            if (headsetColliding)
+            {
+                gameObject.transform.position = lastGoodHeadsetPosition;
+            }
+
+            if (useGravity && carryMomentum)
+            {
+                var device = SteamVR_Controller.Input((int)controllerIndex);
```

Comment:
You can get the controller velocity from the ControllerEvents

https://github.com/thestonefox/SteamVR_Unity_Toolkit/blob/master/Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_ControllerEvents.cs#L278

This is preferable rather than relying on the SteamVR stuff as it uses common abstractions (that may improve compatibility in the future)

You could achieve this with the index by doing something like:

```
var device = VRTK_DeviceFinder.ControllerByIndex(index)
playerPhysics.StartPhysicsFall(-device.GetComponent<VRTK_ControllerEvents>().GetVelocity());
```